From f321f239eb7d28861a16d6f941bd3b756621147c Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Wed, 9 Dec 2015 14:30:21 +0100
Subject: [PATCH 06/28] set the CPU model to kvm64/32 instead of qemu64/32

---
 hw/i386/pc.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/hw/i386/pc.c b/hw/i386/pc.c
index 610050eb4f..e96901435a 100644
--- a/hw/i386/pc.c
+++ b/hw/i386/pc.c
@@ -1151,9 +1151,9 @@ void pc_cpus_init(PCMachineState *pcms)
     /* init CPUs */
     if (machine->cpu_model == NULL) {
 #ifdef TARGET_X86_64
-        machine->cpu_model = "qemu64";
+        machine->cpu_model = "kvm64";
 #else
-        machine->cpu_model = "qemu32";
+        machine->cpu_model = "kvm32";
 #endif
     }
 
-- 
2.11.0

