From 78f096b96d63ed2e7eefd5826c5b8ef62f1244eb Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Tue, 12 Jan 2016 09:09:49 +0100
Subject: [PATCH 16/28] vnc: make x509 imply tls again

---
 ui/vnc.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/ui/vnc.c b/ui/vnc.c
index f30687884b..a345bf0d78 100644
--- a/ui/vnc.c
+++ b/ui/vnc.c
@@ -3881,9 +3881,8 @@ void vnc_display_open(const char *id, Error **errp)
         const char *path;
         bool tls = false, x509 = false, x509verify = false;
         tls  = qemu_opt_get_bool(opts, "tls", false);
-        if (tls) {
-            path = qemu_opt_get(opts, "x509");
-
+        path = qemu_opt_get(opts, "x509");
+        if (tls || path) {
             if (path) {
                 x509 = true;
             } else {
-- 
2.11.0

