From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Thomas Lamprecht <t.lamprecht@proxmox.com>
Date: Tue, 10 Mar 2020 12:55:20 +0100
Subject: [PATCH 23/32] PVE: savevm-async: kick AIO wait on block state write

Signed-off-by: Thomas Lamprecht <t.lamprecht@proxmox.com>
---
 savevm-async.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/savevm-async.c b/savevm-async.c
index 54ceeae26c..393d55af2a 100644
--- a/savevm-async.c
+++ b/savevm-async.c
@@ -158,6 +158,7 @@ static void coroutine_fn block_state_write_entry(void *opaque) {
     BlkRwCo *rwco = opaque;
     rwco->ret = blk_co_pwritev(snap_state.target, rwco->offset, rwco->qiov->size,
                                rwco->qiov, 0);
+    aio_wait_kick();
 }
 
 static ssize_t block_state_writev_buffer(void *opaque, struct iovec *iov,
