From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Reiter <s.reiter@proxmox.com>
Date: Mon, 6 Jul 2020 14:40:13 +0200
Subject: [PATCH] PVE: always set dirty counter for non-incremental backups

Signed-off-by: Stefan Reiter <s.reiter@proxmox.com>
---
 pve-backup.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/pve-backup.c b/pve-backup.c
index 9e767d724c..c108f6a745 100644
--- a/pve-backup.c
+++ b/pve-backup.c
@@ -742,9 +742,13 @@ static void coroutine_fn pvebackup_co_prepare(void *opaque)
                     dirty += bdrv_get_dirty_count(bitmap);
                 }
                 di->bitmap = bitmap;
-            } else if (bitmap != NULL) {
+            } else {
                 dirty += di->size;
-                bdrv_release_dirty_bitmap(bitmap);
+
+                /* after a full backup the old dirty bitmap is invalid anyway */
+                if (bitmap != NULL) {
+                    bdrv_release_dirty_bitmap(bitmap);
+                }
             }
 
             int dev_id = proxmox_backup_co_register_image(pbs, devname, di->size, expect_only_dirty, task->errp);
