From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Reiter <s.reiter@proxmox.com>
Date: Mon, 6 Jul 2020 14:40:14 +0200
Subject: [PATCH] PVE: use proxmox_backup_check_incremental

Signed-off-by: Stefan Reiter <s.reiter@proxmox.com>
Signed-off-by: Thomas Lamprecht <t.lamprecht@proxmox.com>
---
 pve-backup.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/pve-backup.c b/pve-backup.c
index c108f6a745..aa62a1da16 100644
--- a/pve-backup.c
+++ b/pve-backup.c
@@ -734,12 +734,16 @@ static void coroutine_fn pvebackup_co_prepare(void *opaque)
                     if (!bitmap) {
                         goto err;
                     }
-                    /* mark entire bitmap as dirty to make full backup first */
-                    bdrv_set_dirty_bitmap(bitmap, 0, di->size);
-                    dirty += di->size;
                 } else {
-                    expect_only_dirty = true;
+                    expect_only_dirty = proxmox_backup_check_incremental(pbs, devname, di->size) != 0;
+                }
+
+                if (expect_only_dirty) {
                     dirty += bdrv_get_dirty_count(bitmap);
+                } else {
+                    /* mark entire bitmap as dirty to make full backup */
+                    bdrv_set_dirty_bitmap(bitmap, 0, di->size);
+                    dirty += di->size;
                 }
                 di->bitmap = bitmap;
             } else {
