From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Wolfgang Bumiller <w.bumiller@proxmox.com>
Date: Fri, 10 Jul 2020 13:22:35 +0200
Subject: [PATCH] pbs: fix missing crypt and compress parameters

Signed-off-by: Wolfgang Bumiller <w.bumiller@proxmox.com>
---
 pve-backup.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/pve-backup.c b/pve-backup.c
index b0e7b51eef..77eb475563 100644
--- a/pve-backup.c
+++ b/pve-backup.c
@@ -962,6 +962,8 @@ UuidInfo *qmp_backup(
         .backup_file = backup_file,
         .has_password = has_password,
         .password = password,
+        .has_keyfile = has_keyfile,
+        .keyfile = keyfile,
         .has_key_password = has_key_password,
         .key_password = key_password,
         .has_fingerprint = has_fingerprint,
@@ -972,6 +974,10 @@ UuidInfo *qmp_backup(
         .backup_time = backup_time,
         .has_use_dirty_bitmap = has_use_dirty_bitmap,
         .use_dirty_bitmap = use_dirty_bitmap,
+        .has_compress = has_compress,
+        .compress = compress,
+        .has_encrypt = has_encrypt,
+        .encrypt = encrypt,
         .has_format = has_format,
         .format = format,
         .has_config_file = has_config_file,
@@ -980,8 +986,6 @@ UuidInfo *qmp_backup(
         .firewall_file = firewall_file,
         .has_devlist = has_devlist,
         .devlist = devlist,
-        .has_compress = has_compress,
-        .has_encrypt = has_encrypt,
         .has_speed = has_speed,
         .speed = speed,
         .errp = errp,
